### PR TITLE
gh-124265: Remove redundant if statements in traceback.c

### PR DESCRIPTION
**Pull request by rruuaanng**
For the specific reasoning, please refer to https://github.com/python/cpython/issues/124265.
If there are any issues with my reasoning, I will close this PR.

* Issue: gh-124265

### CODE REVIEW

File: Python/traceback.c [hunk 1]
```diff
@@ -980,9 +980,6 @@ dump_traceback(int fd, PyThreadState *tstate, int write_header)
             /* Trampoline frame */
             frame = frame->previous;
         }
-        if (frame == NULL) {
```
Comment (skirpichev) [inline Python/traceback.c:983]: Why, but why do you think that ``frame->previous`` on L981 is not ``NULL``?

Comment (rruuaanng) [reply Python/traceback.c:983]: If frame is part of a chain and there might be a situation where it’s traversed backwards, then at that point in the code https://github.com/python/cpython/blob/7a2d77c903f29d7ea08b870b8e3fa2130f667a59/Python/traceback.c#L975
It should break out. That’s what I think.
I’m not sure if my description is accurate. it’s just how I imagine the situation.

**Comment by skirpichev** [comment]
Hmm, failures seems to be related to the pr.

File: Python/traceback.c (hunk 1 above)
Comment (rruuaanng) [reply Python/traceback.c:983]:
```
static void
dump_traceback(int fd, PyThreadState *tstate, int write_header)
{
    if (write_header) {
        PUTS(fd, "Stack (most recent call first):\n");
    }

    if (tstate_is_freed(tstate)) {
        PUTS(fd, "  <tstate is freed>\n");
        return;
    }

    _PyInterpreterFrame *frame = tstate->current_frame;
    if (frame == NULL) {
        PUTS(fd, "  <no Python frame>\n");
        return;
    }

    unsigned int depth = 0;
    while (1) {
        if (MAX_FRAME_DEPTH <= depth) {
            if (MAX_FRAME_DEPTH < depth) {
                PUTS(fd, "plus ");
                _Py_DumpDecimal(fd, depth);
                PUTS(fd, " frames\n");
            }
            break;
        }
        dump_frame(fd, frame);
        frame = frame->previous;
        if (frame == NULL) {
            break;
        }
        if (frame->owner == FRAME_OWNED_BY_CSTACK) {
            /* Trampoline frame */
            frame = frame->previous;
        }
        // if (frame == NULL) {
        //     break;
        // }
        /* Can't have more than one shim frame in a row */
        assert(frame->owner != FRAME_OWNED_BY_CSTACK);
        depth++;
    }
}
```
I hid the targets I needed to modify by commenting them out.
https://github.com/python/cpython/blob/7a2d77c903f29d7ea08b870b8e3fa2130f667a59/Python/traceback.c#L983-L985

Then run `PCbuild/build.bat`, I then ran the tests using `./python Lib\test\test_traceback.py`.
It seemed to run well as it passed all 347 tests in about 8.3 seconds.
```
PS E:\code\cc\cpython\alpha\cpython-main> ./python .\Lib\test\test_traceback.py
Running Debug|x64 interpreter...
...........................................................................................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 347 tests in 8.303s

OK
```
But I can't understand why it fails on CI. Maybe I ran the test incorrectly.

**Comment by rruuaanng** [comment]
> Hmm, failures seems to be related to the pr.

Well, maybe my guess is wrong.

**Comment by rruuaanng** [comment]
I ran the test file which gave an error on ci, but the result seemed to be correct.
```
PS E:\code\cc\cpython\alpha\cpython-main> ./python .\Lib\test\test_faulthandler.py
Running Debug|x64 interpreter...
......s...s.....s.............sssss.ss..s...s
----------------------------------------------------------------------
Ran 45 tests in 18.992s

OK (skipped=12)
```

**Comment by skirpichev** [comment]
> I ran the test file which gave an error on ci, but the result seemed to be correct.

I got locally same result as for hypothesis tests, maybe it depends on the system:
```
$ ./python -m test test_faulthandler
Using random seed: 1615341603
0:00:00 load avg: 2.46 Run 1 test sequentially in a single process
0:00:00 load avg: 2.46 [1/1] test_faulthandler
test test_faulthandler failed -- multiple errors occurred; run in verbose mode for details
test_faulthandler failed (16 failures)

== Tests result: FAILURE ==

1 test failed:
    test_faulthandler

Total duration: 7.0 sec
Total tests: run=45 failures=16 skipped=5
Total test files: run=1/1 failed=1
Result: FAILURE
```

Sorry, but your patch clearly triggers CI failures.

**Pull request closed without merging**